lib/rouge/formatters/html_linewise.rb

```diff
@@ -6,23 +6,18 @@ module Formatters
     class HTMLLinewise < Formatter
       def initialize(formatter, opts={})
         @formatter = formatter
+        @tag_name = opts.fetch(:tag_name, 'div')
         @class_format = opts.fetch(:class, 'line-%i')
       end
 
       def stream(tokens, &b)
-        token_lines(tokens) do |line|
-          yield "<div class=#{next_line_class}>"
-          line.each do |tok, val|
-            yield @formatter.span(tok, val)
-          end
-          yield '</div>'
+        lineno = 0
+        token_lines(tokens) do |line_tokens|
+          yield %(<#{@tag_name} class="#{sprintf @class_format, lineno += 1}">)
+          @formatter.stream(line_tokens) {|formatted| yield formatted }
+          yield %(\n</#{@tag_name}>)
```

Comment:
Why introduce a newline here?

Reply:
Because we need content at the end of the line to stay the end of the line so we can do end of line manipulation without having to step over the (artificial) closing tag. This is important for postprocessors like Asciidoctor that need to add and remove content from the end of the line. Without this, I have to guess why there is an extra closing tag and it becomes very fragile.

Reply:
For the record, Pygments honors this contract.

Reply:
I've tested this with a Jekyll installation and am not seeing any regressions. Are you OK to resolve this @dblessing?

Reply:
Gitlab uses a custom formatter iirc, and they only rely on the behavior of `#token_lines`.

Reply:
(this is the approach we've been encouraging for larger projects, as formatters are relatively straightforward to write, and it's not reasonable to try to cover *everyone's* needs with a plain old option hash. also it makes those projects more robust to breaking changes in the formatter stack.)

Reply:
Thanks for the clarification, @jneen! Agreed that encouraging bespoke formatters is a better alternative. Having better understood this, I don't think this will break things but it would nevertheless be a good idea to encourage Jekyll to adopt that approach so that the risk of that kind of thing occurring is minimised.